httomo/yaml_checker.py

```diff
@@ -216,114 +130,208 @@ def check_methods_exist_in_templates(conf: PipelineConfig) -> bool:
+            return False
+    return True
+
+
+def check_side_out_matches_ref_arg(conf: PipelineConfig) -> bool:
+    """ Check reference name exists """
+    ref_ids = {k: v for m in conf for k, v in get_valid_ref_ids(m).items()}
+    for k, v in ref_ids.items():
+        (ref_id, side_str, ref_arg) = get_ref_split(v)
+        side_dicts = [
+            m.get(side_str)
+            for m in conf
+            if m.get("id") == ref_id and m.get(side_str) is not None
+        ]
+        all_side_out = {k: v for d in side_dicts for k, v in d.items()}
+        if ref_arg not in all_side_out.values():
+            _print_with_colour(
+                f"The reference value: {ref_arg} was not found to have a matching side"
+                f"output value."
+            )
+            return False
+    return True
+
+
+def _get_template_yaml_conf(conf: PipelineConfig) -> PipelineConfig:
+    """Get the pipeline config method dictionaries from template yaml files
+
+    Parameters
+    ----------
+    conf
+       The list of method dictionaries in the pipeline
+        this specifies which yaml template methods to get
+
+    Returns
+    -------
+    list of method dictionaries which is loaded from yaml templates
     """
-    modules: List[str] = []
-    methods: List[MethodConfig] = []
-    for stage in conf:
-        for module in stage:
-            module_name = list(module.keys())[0]
-            modules.append(module_name)
-            methods.append(module[module_name])
+    packages = _get_package_info(conf)
+    template_yaml_files = _get_template_yaml(conf, packages)
+    template_yaml_conf: PipelineConfig = []
+    for f in template_yaml_files:
+        tmp_conf = yaml_loader(f)
+        # make an assumption there is one method inside each template
+        template_yaml_conf.append(tmp_conf[0])
+    return template_yaml_conf
 
+
+def _get_package_info(conf: PipelineConfig) -> List:
+    """
+    Helper function to get packages from module path.
+    """
+    modules = [m["module_path"] for m in conf]
     packages = [
-        m.split(".")[0] + "/" + get_external_package_current_version(m.split(".")[0])
+        m.split(".")[0] + "/" + get_packages_current_version(m.split(".")[0])
         if m.split(".")[0] != "httomo"
         else m.split(".")[0]
+        + "/"
+        + get_packages_current_version(m.split(".")[0], type_p="httomo")
         for m in modules
     ]
-    return modules, methods, packages
+    return packages
 
 
-def _get_yaml_templates(modules: List, methods: List, packages: List) -> List:
+def _get_template_yaml(conf: PipelineConfig, packages: List) -> List:
     """
-    Helper function that fetches YAML template files associated with methods
+    Helper function that fetches template YAML file names associated with methods
     passed.
     """
     parent_dir = os.path.dirname(os.path.abspath("__file__"))
     templates_dir = os.path.join(parent_dir, "templates")
     assert os.path.exists(templates_dir)
```

Comment:
Thanks for replacing the various `assert` statements in this file with printing of a helpful error message to describe the issue before returning a `bool` :slightly_smiling_face: It makes it easier for anything calling the top-level `validate_yaml_config()` function to simply:
- execute that function
- let any failures be printed
- and finally check the `bool` value that was returned

There was something that came to mind when I thought about the launcher python script calling `validate_yaml_config()` and any error handling it would have to do to terminate the program before the job was submitted.

This `assert` looks like it might be the only place now where the `validate_yaml_config()` function can possibly fail and a nice error message _isn't_ printed.

Ie, even if the caller does something like:
```python
try:
    validate_yaml_config(my_yaml)
except AssertionErrror as e:
   print(f"YAML check failed with unexpected error: {e}")
```

and the assertion failed, the `e` would have no error message within to use for printing, because no error message is reported when the error is raised.

Do you think it would make sense to given an error message in case the assertion fails, so then a caller can use a try/except as above to then handle the error and avoid displaying the stacktrace to the user?

Ie, something like:
```python
assert os.path.exists(templates_dir), "Dev error: expected YAML templates dir to exist"
```

which could then be nicely handled by the caller in the unlikely event that the templates directory doesn't exist for whatever reason?

Reply:
Thank you for reviewing. For the error messages, I had been following the format which Naman had set up. This assert statement does not take into account logging and error messages. This was a short term solution under time constraint to halt the later run if there was an error inside the yaml file. It definitely is sensible to include clearer logging capability, as had previously discussed, but this hasn't been included in this branch.

Reply:
> It definitely is sensible to include clearer logging capability, as had previously discussed, but this hasn't been included in this branch.

To be clear, I'm not suggesting adding logging capabilities or anything, I'm suggesting to add a message to the `assert` that it should return if it fails, a one-line change like:
```suggestion
    assert os.path.exists(templates_dir), "Dev error: expected YAML templates dir to exist"
```

so then callers of the function can report what went wrong without displaying a nasty stacktrace to the user.

Reply:
Thanks for clarifying, added in [f53709a](https://github.com/DiamondLightSource/httomo/pull/201/commits/f53709ac6350898dcfe874961241d98e3a722f84)